processor/servicegraphprocessor/processor.go

```diff
@@ -0,0 +1,504 @@
+	"time"
+
+	"go.opencensus.io/stats"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/servicegraphprocessor/internal/store"
+)
+
+const (
+	metricKeySeparator = string(byte(0))
+)
+
+var (
+	maxDuration   = time.Duration(math.MaxInt64)
+	maxDurationMs = durationToMillis(maxDuration)
+
+	defaultLatencyHistogramBucketsMs = []float64{
+		2, 4, 6, 8, 10, 50, 100, 200, 400, 800, 1000, 1400, 2000, 5000, 10_000, 15_000, maxDurationMs,
+	}
+)
+
+type metricSeries struct {
+	dimensions  pcommon.Map
+	lastUpdated int64 // Used to remove stale series
+}
+
+var _ component.TracesProcessor = (*processor)(nil)
+
+type processor struct {
+	config          *Config
+	logger          *zap.Logger
+	nextConsumer    consumer.Traces
+	metricsExporter consumer.Metrics
+
+	store store.Store
+
+	startTime time.Time
+
+	seriesMutex                    sync.Mutex
+	reqTotal                       map[string]int64
+	reqFailedTotal                 map[string]int64
+	reqDurationSecondsSum          map[string]float64
+	reqDurationSecondsCount        map[string]uint64
+	reqDurationBounds              []float64
+	reqDurationSecondsBucketCounts map[string][]uint64
+
+	keyToMetric map[string]metricSeries
+
+	shutdownCh chan interface{}
+}
+
+func newProcessor(logger *zap.Logger, config config.Processor, nextConsumer consumer.Traces) *processor {
+	logger.Info("Building servicegraphsprocessor")
+
+	pConfig := config.(*Config)
+
+	bounds := defaultLatencyHistogramBucketsMs
+	if pConfig.LatencyHistogramBuckets != nil {
+		bounds = mapDurationsToMillis(pConfig.LatencyHistogramBuckets)
+
+		// "Catch-all" bucket.
+		// TODO: Revisit catch-all bucket, it's very problematic for Prometheus
+		//  See https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/2838
+		if bounds[len(bounds)-1] != maxDurationMs {
+			bounds = append(bounds, maxDurationMs)
+		}
+	}
+
+	p := &processor{
+		config:                         pConfig,
+		logger:                         logger,
+		nextConsumer:                   nextConsumer,
+		startTime:                      time.Now(),
+		reqTotal:                       make(map[string]int64),
+		reqFailedTotal:                 make(map[string]int64),
+		reqDurationSecondsSum:          make(map[string]float64),
+		reqDurationSecondsCount:        make(map[string]uint64),
+		reqDurationBounds:              bounds,
+		reqDurationSecondsBucketCounts: make(map[string][]uint64),
+		keyToMetric:                    make(map[string]metricSeries),
+		shutdownCh:                     make(chan interface{}),
+	}
+
+	return p
+}
+
+func (p *processor) Start(_ context.Context, host component.Host) error {
+	p.logger.Info("Starting servicegraphsprocessor")
```

Comment:
Similar to a previous comment: I think the pipeline builder logs this already.

Reply:
Removed!